# Chapter 1 — Introduction

## Introduction to Machine Learning with Python

**Book:** Introduction to Machine Learning with Python  
**Authors:** Andreas C. Müller & Sarah Guido

## 1. Chapter Overview

Chapter ini memberikan pengenalan terhadap konsep dasar
Machine Learning dan penggunaan Python untuk membangun
model Machine Learning.

Chapter ini juga memperkenalkan scikit-learn sebagai salah
satu library utama untuk implementasi Machine Learning di Python.

Pada bagian akhir chapter, dilakukan contoh sederhana
menggunakan dataset Iris dan algoritma K-Nearest Neighbors.

## 2. Learning Objectives

Setelah mempelajari chapter ini, tujuan pembelajaran yang ingin
dicapai adalah:

1. Memahami konsep dasar Machine Learning.
2. Memahami jenis-jenis permasalahan Machine Learning.
3. Mengenal supervised dan unsupervised learning.
4. Memahami konsep training dan testing data.
5. Mengenal scikit-learn.
6. Mengimplementasikan model K-Nearest Neighbors.
7. Melakukan prediksi menggunakan model Machine Learning.
8. Mengevaluasi performa model.

## 3. Introduction to Machine Learning

### 3.1 What is Machine Learning?

Machine Learning merupakan pendekatan yang memungkinkan
komputer mempelajari pola dari data untuk menghasilkan prediksi
atau keputusan tanpa harus menentukan seluruh aturan secara
eksplisit.

Dalam Machine Learning, model mempelajari hubungan atau pola
yang terdapat pada data training dan kemudian menggunakan pola
tersebut untuk membuat prediksi terhadap data baru.

### 3.2 Traditional Programming vs Machine Learning

Pada pemrograman tradisional, programmer menentukan aturan
yang digunakan untuk menghasilkan output dari suatu input.

Pada Machine Learning, data dan hasil yang diharapkan digunakan
untuk melatih model sehingga model dapat mempelajari pola yang
terdapat pada data tersebut.

## 4. Supervised Learning

Supervised Learning merupakan jenis Machine Learning yang
menggunakan data yang memiliki target atau label.

Model mempelajari hubungan antara input dan target berdasarkan
data training. Setelah proses training selesai, model dapat
digunakan untuk memprediksi target dari data yang belum pernah
dilihat sebelumnya.

Dua permasalahan utama dalam supervised learning adalah:

1. Classification
2. Regression

### Classification

Classification digunakan ketika target yang ingin diprediksi
berupa kategori.

Contohnya adalah memprediksi apakah sebuah email termasuk
spam atau bukan spam.

### Regression

Regression digunakan ketika target yang ingin diprediksi
berupa nilai numerik.

Contohnya adalah memprediksi harga sebuah rumah.

## 5. Unsupervised Learning

Unsupervised Learning merupakan pendekatan Machine Learning
yang menggunakan data tanpa target atau label.

Tujuannya adalah menemukan pola atau struktur yang terdapat
di dalam data.

Salah satu contoh permasalahan unsupervised learning adalah
clustering, yaitu mengelompokkan data berdasarkan kemiripan
karakteristiknya.

## 6. Machine Learning with Python

Python merupakan salah satu bahasa pemrograman yang banyak
digunakan dalam pengembangan Machine Learning karena memiliki
berbagai library untuk pengolahan data, visualisasi, dan
pembangunan model.

Salah satu library utama yang digunakan dalam buku ini adalah
scikit-learn.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import sklearn

### Explanation

Library yang digunakan memiliki fungsi yang berbeda:

- NumPy digunakan untuk operasi numerik dan array.
- pandas digunakan untuk pengolahan dan analisis data.
- matplotlib digunakan untuk membuat visualisasi.
- scikit-learn digunakan untuk implementasi algoritma
  Machine Learning.

## 7. The Iris Dataset

Dataset Iris merupakan dataset yang digunakan untuk
memperkenalkan proses pembangunan model Machine Learning.

Dataset ini berisi karakteristik bunga Iris yang dapat
digunakan untuk melakukan klasifikasi berdasarkan beberapa
fitur bunga.

In [2]:
from sklearn.datasets import load_iris

iris_dataset = load_iris()

In [3]:
print(iris_dataset.keys())

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


### Explanation

Objek `iris_dataset` berisi informasi yang berkaitan dengan
dataset Iris.

Beberapa informasi penting di dalamnya adalah:

- `data` berisi nilai fitur dari setiap sampel.
- `target` berisi label atau kelas setiap sampel.
- `target_names` berisi nama kelas.
- `feature_names` berisi nama fitur yang digunakan.
- `DESCR` berisi deskripsi dataset.

In [4]:
print(iris_dataset['target_names'])

['setosa' 'versicolor' 'virginica']


In [5]:
print(iris_dataset['feature_names'])

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [6]:
print(iris_dataset['data'].shape)

(150, 4)


In [7]:
print(iris_dataset['target'].shape)

(150,)


### Features and Target

Feature merupakan variabel atau karakteristik yang digunakan
sebagai input untuk model Machine Learning.

Pada dataset Iris, fitur yang digunakan berkaitan dengan
karakteristik bunga.

Target merupakan nilai yang ingin diprediksi oleh model.
Pada kasus ini, target menunjukkan kelas atau jenis bunga Iris.

## 8. Training and Testing Data

Sebelum membangun model, dataset perlu dibagi menjadi training
data dan testing data.

Training data digunakan untuk melatih model agar mempelajari
pola dari data.

Testing data digunakan untuk mengevaluasi kemampuan model
terhadap data yang tidak digunakan selama proses training.

In [8]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    iris_dataset['data'],
    iris_dataset['target'],
    random_state=0
)

### Explanation

Variabel `X` digunakan untuk menyimpan fitur atau input,
sedangkan `y` digunakan untuk menyimpan target atau label.

Setelah dilakukan train-test split:

- `X_train` berisi fitur untuk training.
- `X_test` berisi fitur untuk testing.
- `y_train` berisi target untuk training.
- `y_test` berisi target untuk testing.

Parameter `random_state=0` digunakan agar pembagian data dapat
direproduksi dan menghasilkan pembagian yang sama ketika kode
dijalankan kembali.

## 9. Building a K-Nearest Neighbors Model

K-Nearest Neighbors atau KNN merupakan algoritma supervised
learning yang melakukan prediksi berdasarkan data training
yang memiliki jarak paling dekat dengan data yang akan
diprediksi.

Parameter `n_neighbors` menentukan jumlah tetangga yang
digunakan oleh model.

In [9]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=1)

In [10]:
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=1)

### Explanation

Method `fit()` digunakan untuk melatih model menggunakan
training data.

Pada kode tersebut:

- `X_train` merupakan fitur training.
- `y_train` merupakan target atau label training.

Setelah proses `fit()` selesai, model dapat digunakan untuk
melakukan prediksi terhadap data baru.

## 10. Making Predictions

In [11]:
import numpy as np

X_new = np.array([[5, 2.9, 1, 0.2]])

prediction = knn.predict(X_new)

print("Prediction:", prediction)
print("Predicted target name:",
      iris_dataset['target_names'][prediction])

Prediction: [0]
Predicted target name: ['setosa']


### Explanation

`X_new` merupakan data baru yang ingin diklasifikasikan.

Method `predict()` digunakan untuk menghasilkan prediksi
berdasarkan pola yang telah dipelajari model.

Hasil prediksi berupa label numerik yang kemudian dapat
dikonversi menjadi nama kelas menggunakan `target_names`.

## 11. Evaluating the Model

In [12]:
test_score = knn.score(X_test, y_test)

print("Test set score:", test_score)

Test set score: 0.9736842105263158


### Explanation

Method `score()` digunakan untuk mengukur performa model
terhadap testing data.

Pada classification, score yang digunakan oleh
`KNeighborsClassifier` adalah accuracy.

Accuracy menunjukkan proporsi data testing yang berhasil
diklasifikasikan dengan benar.

## 12. Results and Analysis

Model K-Nearest Neighbors menghasilkan test set score sebesar
0.9737 atau sekitar 97,37%.

Hasil tersebut menunjukkan bahwa sebagian besar data pada
testing set berhasil diklasifikasikan dengan benar.

Namun, nilai tersebut diperoleh dari satu pembagian training
dan testing data dengan `random_state=0`. Oleh karena itu,
hasil ini tidak dapat langsung dianggap sebagai performa model
pada seluruh kemungkinan data.

# 13. Chapter Summary

Pada chapter ini dipelajari konsep dasar Machine Learning,
jenis permasalahan supervised dan unsupervised learning, serta
penggunaan Python dan scikit-learn untuk membangun model
Machine Learning.

Implementasi dilakukan menggunakan dataset Iris dan algoritma
K-Nearest Neighbors. Proses dimulai dari memuat dataset,
memahami fitur dan target, membagi data menjadi training dan
testing data, membangun model, melakukan training, membuat
prediksi, dan mengevaluasi performa model.

Dari contoh tersebut dapat dipahami bahwa proses Machine
Learning tidak hanya terdiri dari pembuatan model, tetapi juga
melibatkan persiapan data, training, prediction, dan evaluation.

# 14. Key Takeaways

- Machine Learning memungkinkan komputer mempelajari pola dari data.
- Supervised learning menggunakan data yang memiliki target.
- Classification digunakan untuk memprediksi kategori.
- Regression digunakan untuk memprediksi nilai numerik.
- Dataset dapat dibagi menjadi training dan testing data.
- Scikit-learn menyediakan berbagai algoritma Machine Learning.
- KNN melakukan prediksi berdasarkan tetangga terdekat.
- Model perlu dievaluasi menggunakan testing data.